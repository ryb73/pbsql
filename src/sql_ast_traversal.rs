pub mod ast_views;
pub mod helpers;
pub mod traverser;
