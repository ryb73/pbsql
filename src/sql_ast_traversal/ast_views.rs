use sqlparser::ast::{
    Assignment, ColumnDef, Expr, FileFormat, HiveDistributionStyle, HiveFormat, Ident,
    MysqlInsertPriority, ObjectName, ObjectType, OnCommit, OnInsert, OrderByExpr, Query,
    SelectItem, SetExpr, SetOperator, SetQuantifier, SqlOption, SqliteOnConflict, Statement,
    TableAlias, TableConstraint, TableFactor, TableWithJoins,
};

#[derive(Debug)]
pub struct DropStatementViewMutable<'a> {
    pub cascade: &'a mut bool,
    pub if_exists: &'a mut bool,
    pub names: &'a mut Vec<ObjectName>,
    pub object_type: &'a mut ObjectType,
    pub purge: &'a mut bool,
    pub restrict: &'a mut bool,
    pub temporary: &'a mut bool,
}

impl<'a> TryFrom<&'a mut Statement> for DropStatementViewMutable<'a> {
    type Error = String;

    fn try_from(statement: &'a mut Statement) -> Result<Self, Self::Error> {
        match statement {
            Statement::Drop {
                cascade,
                if_exists,
                names,
                object_type,
                purge,
                restrict,
                temporary,
            } => Ok(DropStatementViewMutable {
                cascade,
                if_exists,
                names,
                object_type,
                purge,
                restrict,
                temporary,
            }),
            _ => Err("Expected a Drop statement".to_string()),
        }
    }
}

#[derive(Debug)]
pub struct UpdateStatementViewMutable<'a> {
    pub assignments: &'a mut Vec<Assignment>,
    pub from: &'a mut Option<TableWithJoins>,
    pub returning: &'a mut Option<Vec<SelectItem>>,
    pub selection: &'a mut Option<Expr>,
    pub table: &'a mut TableWithJoins,
}

impl<'a> TryFrom<&'a mut Statement> for UpdateStatementViewMutable<'a> {
    type Error = String;

    fn try_from(statement: &'a mut Statement) -> Result<Self, Self::Error> {
        match statement {
            Statement::Update {
                assignments,
                from,
                returning,
                selection,
                table,
            } => Ok(UpdateStatementViewMutable {
                assignments,
                from,
                returning,
                selection,
                table,
            }),
            _ => Err("Expected an Update statement".to_string()),
        }
    }
}

#[derive(Debug)]
pub struct CreateIndexStatementViewMutable<'a> {
    pub columns: &'a mut Vec<OrderByExpr>,
    pub name: &'a mut Option<ObjectName>,
    pub table_name: &'a mut ObjectName,
    pub unique: &'a mut bool,
    pub if_not_exists: &'a mut bool,
    pub nulls_distinct: &'a mut Option<bool>,
    pub concurrently: &'a mut bool,
    pub include: &'a mut Vec<Ident>,
    pub predicate: &'a mut Option<Expr>,
    pub using: &'a mut Option<Ident>,
}

impl<'a> TryFrom<&'a mut Statement> for CreateIndexStatementViewMutable<'a> {
    type Error = String;

    fn try_from(statement: &'a mut Statement) -> Result<Self, Self::Error> {
        match statement {
            Statement::CreateIndex {
                columns,
                name,
                table_name,
                unique,
                if_not_exists,
                nulls_distinct,
                concurrently,
                include,
                predicate,
                using,
            } => Ok(CreateIndexStatementViewMutable {
                columns,
                name,
                table_name,
                unique,
                if_not_exists,
                nulls_distinct,
                concurrently,
                include,
                predicate,
                using,
            }),
            _ => Err("Expected a CreateIndex statement".to_string()),
        }
    }
}

#[derive(Debug)]
pub struct CreateTableStatementViewMutable<'a> {
    pub if_not_exists: &'a mut bool,
    pub name: &'a mut ObjectName,
    pub columns: &'a mut Vec<ColumnDef>,
    pub auto_increment_offset: &'a mut Option<u32>,
    pub clone: &'a mut Option<ObjectName>,
    pub cluster_by: &'a mut Option<Vec<Ident>>,
    pub collation: &'a mut Option<String>,
    pub comment: &'a mut Option<String>,
    pub constraints: &'a mut Vec<TableConstraint>,
    pub default_charset: &'a mut Option<String>,
    pub engine: &'a mut Option<String>,
    pub external: &'a mut bool,
    pub file_format: &'a mut Option<FileFormat>,
    pub global: &'a mut Option<bool>,
    pub hive_distribution: &'a mut HiveDistributionStyle,
    pub hive_formats: &'a mut Option<HiveFormat>,
    pub like: &'a mut Option<ObjectName>,
    pub location: &'a mut Option<String>,
    pub on_cluster: &'a mut Option<String>,
    pub on_commit: &'a mut Option<OnCommit>,
    pub options: &'a mut Option<Vec<SqlOption>>,
    pub or_replace: &'a mut bool,
    pub order_by: &'a mut Option<Vec<Ident>>,
    pub partition_by: &'a mut Option<Box<Expr>>,
    pub query: &'a mut Option<Box<Query>>,
    pub strict: &'a mut bool,
    pub table_properties: &'a mut Vec<SqlOption>,
    pub temporary: &'a mut bool,
    pub transient: &'a mut bool,
    pub with_options: &'a mut Vec<SqlOption>,
    pub without_rowid: &'a mut bool,
}

impl<'a> TryFrom<&'a mut Statement> for CreateTableStatementViewMutable<'a> {
    type Error = String;

    fn try_from(statement: &'a mut Statement) -> Result<Self, Self::Error> {
        match statement {
            Statement::CreateTable {
                if_not_exists,
                name,
                columns,
                auto_increment_offset,
                clone,
                cluster_by,
                collation,
                comment,
                constraints,
                default_charset,
                engine,
                external,
                file_format,
                global,
                hive_distribution,
                hive_formats,
                like,
                location,
                on_cluster,
                on_commit,
                options,
                or_replace,
                order_by,
                partition_by,
                query,
                strict,
                table_properties,
                temporary,
                transient,
                with_options,
                without_rowid,
            } => Ok(CreateTableStatementViewMutable {
                if_not_exists,
                name,
                columns,
                auto_increment_offset,
                clone,
                cluster_by,
                collation,
                comment,
                constraints,
                default_charset,
                engine,
                external,
                file_format,
                global,
                hive_distribution,
                hive_formats,
                like,
                location,
                on_cluster,
                on_commit,
                options,
                or_replace,
                order_by,
                partition_by,
                query,
                strict,
                table_properties,
                temporary,
                transient,
                with_options,
                without_rowid,
            }),
            _ => Err("Expected a CreateTable statement".to_string()),
        }
    }
}

#[derive(Debug)]
pub struct InsertStatementViewMutable<'a> {
    pub after_columns: &'a mut Vec<Ident>,
    pub columns: &'a mut Vec<Ident>,
    pub ignore: &'a mut bool,
    pub into: &'a mut bool,
    pub on: &'a mut Option<OnInsert>,
    pub or: &'a mut Option<SqliteOnConflict>,
    pub overwrite: &'a mut bool,
    pub partitioned: &'a mut Option<Vec<Expr>>,
    pub priority: &'a mut Option<MysqlInsertPriority>,
    pub replace_into: &'a mut bool,
    pub returning: &'a mut Option<Vec<SelectItem>>,
    pub source: &'a mut Option<Box<Query>>,
    pub table_alias: &'a mut Option<Ident>,
    pub table_name: &'a mut ObjectName,
    pub table: &'a mut bool,
}

impl<'a> TryFrom<&'a mut Statement> for InsertStatementViewMutable<'a> {
    type Error = String;

    fn try_from(statement: &'a mut Statement) -> Result<Self, Self::Error> {
        match statement {
            Statement::Insert {
                after_columns,
                columns,
                ignore,
                into,
                on,
                or,
                overwrite,
                partitioned,
                priority,
                replace_into,
                returning,
                source,
                table_alias,
                table_name,
                table,
            } => Ok(InsertStatementViewMutable {
                after_columns,
                columns,
                ignore,
                into,
                on,
                or,
                overwrite,
                partitioned,
                priority,
                replace_into,
                returning,
                source,
                table_alias,
                table_name,
                table,
            }),
            _ => Err("Expected an Insert statement".to_string()),
        }
    }
}

#[derive(Debug)]
pub struct SetOperationViewMutable<'a> {
    pub left: &'a mut Box<SetExpr>,
    pub op: &'a mut SetOperator,
    pub right: &'a mut Box<SetExpr>,
    pub set_quantifier: &'a mut SetQuantifier,
}

impl<'a> TryFrom<&'a mut SetExpr> for SetOperationViewMutable<'a> {
    type Error = String;

    fn try_from(set_expr: &'a mut SetExpr) -> Result<Self, Self::Error> {
        match set_expr {
            SetExpr::SetOperation {
                left,
                op,
                right,
                set_quantifier,
            } => Ok(SetOperationViewMutable {
                left,
                op,
                right,
                set_quantifier,
            }),
            _ => Err("Expected a SetOperation".to_string()),
        }
    }
}

#[derive(Debug)]
pub struct InSubqueryExprViewMutable<'a> {
    pub expr: &'a mut Box<Expr>,
    pub negated: &'a mut bool,
    pub subquery: &'a mut Box<Query>,
}

impl<'a> TryFrom<&'a mut Expr> for InSubqueryExprViewMutable<'a> {
    type Error = String;

    fn try_from(expr: &'a mut Expr) -> Result<Self, Self::Error> {
        match expr {
            Expr::InSubquery {
                expr,
                negated,
                subquery,
            } => Ok(InSubqueryExprViewMutable {
                expr,
                negated,
                subquery,
            }),
            _ => Err("Expected an InSubquery expression".to_string()),
        }
    }
}

#[derive(Debug)]
pub struct DerivedTableFactorViewMutable<'a> {
    pub alias: &'a mut Option<TableAlias>,
    pub lateral: &'a mut bool,
    pub subquery: &'a mut Box<Query>,
}

impl<'a> TryFrom<&'a mut TableFactor> for DerivedTableFactorViewMutable<'a> {
    type Error = String;

    fn try_from(table_factor: &'a mut TableFactor) -> Result<Self, Self::Error> {
        match table_factor {
            TableFactor::Derived {
                alias,
                lateral,
                subquery,
            } => Ok(DerivedTableFactorViewMutable {
                alias,
                lateral,
                subquery,
            }),
            _ => Err("Expected a DerivedTableFactor".to_string()),
        }
    }
}
