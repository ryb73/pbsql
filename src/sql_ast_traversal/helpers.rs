use sqlparser::ast::Ident;

pub fn extract_unary_identifier(
    identifiers: &Vec<Ident>,
    object_type: &str,
) -> Result<String, String> {
    if identifiers.len() != 1 {
        return Err(format!(
            "Expected 1 identifiers for the {} name, got: {:?}",
            object_type,
            identifiers
                .iter()
                .map(|x| x.to_string())
                .collect::<Vec<_>>()
        ));
    }

    Ok(identifiers[0].value.clone())
}

pub fn extract_binary_identifiers(
    identifiers: &Vec<Ident>,
    object_type: &str,
) -> Result<(String, String), String> {
    if identifiers.len() != 2 {
        return Err(format!(
            "Expected 2 identifiers for the {} name, got: {:?}",
            object_type,
            identifiers
                .iter()
                .map(|x| x.to_string())
                .collect::<Vec<_>>()
        ));
    }

    Ok((identifiers[0].value.clone(), identifiers[1].value.clone()))
}
