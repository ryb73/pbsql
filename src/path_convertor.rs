use super::sql_ast_traversal::{
    ast_views::{
        CreateIndexStatementViewMutable, CreateTableStatementViewMutable, DropStatementViewMutable,
        InsertStatementViewMutable, SetOperationViewMutable, TableFactorDerivedViewMut,
        TableFactorTableViewMut, UpdateStatementViewMutable,
    },
    helpers::{extract_binary_identifiers, extract_unary_identifier},
    traverser::VisitResult,
};
use crate::{
    sql_ast_traversal::traverser::SqlAstTraverser, VALUES_TABLE_INDEX_PREFIX, VALUES_TABLE_NAME,
};
use sqlparser::ast::{
    self, Assignment, Function, FunctionArgExpr, HiveDistributionStyle, HiveFormat, Ident,
    JoinConstraint, ObjectName, ObjectType, OnConflict, TableAlias, Value,
};
use std::collections::HashMap;
use typed_path::Utf8UnixPathBuf;

pub type DatabaseNamesByPath = HashMap<String, String>;

type TableReference = String;
type ReplacementIdentifiers = Vec<Ident>;
type ReferencableTables = HashMap<TableReference, ReplacementIdentifiers>;

#[derive(Debug)]
struct Scope {
    referencable_tables: ReferencableTables,
}

impl Scope {
    fn new() -> Self {
        Scope {
            referencable_tables: HashMap::new(),
        }
    }
}

pub struct PathConvertor {
    pub database_names: DatabaseNamesByPath,
    scopes: Vec<Scope>,
}

impl PathConvertor {
    pub fn new() -> Self {
        PathConvertor {
            database_names: HashMap::new(),
            scopes: vec![Scope::new()],
        }
    }
}

impl SqlAstTraverser for PathConvertor {
    fn post_visit_drop(&mut self, drop: &mut DropStatementViewMutable) -> VisitResult {
        let DropStatementViewMutable {
            cascade: _,
            if_exists: _,
            names,
            object_type: _,
            purge: _,
            restrict: _,
            temporary: _,
        } = drop;

        let new_names = names
            .iter()
            .map(|name| {
                let db_reference = convert_path_to_database(name, &mut self.database_names)?;
                Ok(ObjectName(get_qualified_values_table_identifiers(
                    &db_reference,
                )))
            })
            .collect::<Result<Vec<_>, String>>()?;

        **names = new_names;

        Ok(())
    }

    fn pre_visit_update(&mut self, _update: &mut UpdateStatementViewMutable) -> VisitResult {
        self.scopes.push(Scope::new());
        Ok(())
    }

    fn post_visit_update(&mut self, _update: &mut UpdateStatementViewMutable) -> VisitResult {
        self.scopes.pop();
        Ok(())
    }

    fn post_visit_assignment(&mut self, assignment: &mut Assignment) -> VisitResult {
        let Assignment { id, value: _ } = assignment;
        extract_unary_identifier(id, "column")?;
        Ok(())
    }

    fn pre_visit_create_table(
        &mut self,
        create_table: &mut CreateTableStatementViewMutable,
    ) -> VisitResult {
        let CreateTableStatementViewMutable {
            if_not_exists: _,
            name: _,
            columns: _,
            auto_increment_offset: _,
            clone,
            cluster_by,
            collation: _,
            comment,
            constraints: _,
            default_charset: _,
            engine,
            external,
            file_format,
            global,
            hive_distribution,
            hive_formats,
            like,
            location,
            on_cluster,
            on_commit: _,
            options: _,
            or_replace: _,
            order_by,
            partition_by: _,
            query: _,
            strict: _,
            table_properties: _,
            temporary,
            transient,
            with_options: _,
            without_rowid,
        } = create_table;

        if clone.is_some() {
            return Err("not implemented: CreateTable::clone".to_string());
        }

        if cluster_by.is_some() {
            return Err("not implemented: CreateTable::cluster_by".to_string());
        }

        if comment.is_some() {
            return Err("not implemented: CreateTable::comment".to_string());
        }

        if engine.is_some() {
            return Err("not implemented: CreateTable::engine".to_string());
        }

        if **external {
            return Err("not implemented: CreateTable::external".to_string());
        }

        if file_format.is_some() {
            return Err("not implemented: CreateTable::file_format".to_string());
        }

        if global.is_some() {
            return Err("not implemented: CreateTable::global".to_string());
        }

        if **hive_distribution != HiveDistributionStyle::NONE {
            return Err("not implemented: CreateTable::hive_distribution".to_string());
        }

        if let Some(HiveFormat {
            location,
            row_format,
            storage,
        }) = hive_formats
        {
            if location.is_some() || row_format.is_some() || storage.is_some() {
                return Err("not implemented: CreateTable::hive_formats".to_string());
            }
        }

        if like.is_some() {
            return Err("not implemented: CreateTable::like".to_string());
        }

        if location.is_some() {
            return Err("not implemented: CreateTable::location".to_string());
        }

        if on_cluster.is_some() {
            return Err("not implemented: CreateTable::on_cluster".to_string());
        }

        if order_by.is_some() {
            return Err("not implemented: CreateTable::order_by".to_string());
        }

        if **temporary {
            return Err("not implemented: CreateTable::temporary".to_string());
        }

        if **transient {
            return Err("not implemented CreateTable::transient".to_string());
        }

        if **without_rowid {
            return Err("not implemented: CreateTable::without_rowid".to_string());
        }

        Ok(())
    }

    fn post_visit_create_table(
        &mut self,
        create_table: &mut CreateTableStatementViewMutable,
    ) -> VisitResult {
        let db_reference = convert_path_to_database(create_table.name, &mut self.database_names)?;

        *create_table.name = ObjectName(get_qualified_values_table_identifiers(&db_reference));

        Ok(())
    }

    fn pre_visit_create_index(
        &mut self,
        create_index: &mut CreateIndexStatementViewMutable,
    ) -> VisitResult {
        let CreateIndexStatementViewMutable {
            columns: _,
            name,
            table_name: _,
            unique: _,
            if_not_exists: _,
            nulls_distinct: _,
            concurrently,
            include,
            predicate: _,
            using,
        } = create_index;

        if name.is_none() {
            return Err("Index name is required".to_string());
        }

        if **concurrently {
            return Err("not implemented: CreateIndex::concurrently".to_string());
        }

        if !include.is_empty() {
            return Err("not implemented: CreateIndex::include".to_string());
        }

        if using.is_some() {
            return Err("not implemented: CreateIndex::using".to_string());
        }

        Ok(())
    }

    fn post_visit_create_index(
        &mut self,
        create_index: &mut CreateIndexStatementViewMutable,
    ) -> VisitResult {
        let ObjectName(name_identifiers) = create_index.name.as_ref().unwrap();

        let index_name = extract_unary_identifier(name_identifiers, "index")?;

        let translated_db_name =
            convert_path_to_database(&create_index.table_name, &mut self.database_names)?;

        *create_index.name = Some(ObjectName(vec![
            Ident::new(translated_db_name),
            Ident::new(format!("{}{}", VALUES_TABLE_INDEX_PREFIX, index_name)),
        ]));
        *create_index.table_name = ObjectName(vec![Ident::new(VALUES_TABLE_NAME)]);

        Ok(())
    }

    fn pre_visit_insert(&mut self, insert: &mut InsertStatementViewMutable) -> VisitResult {
        let InsertStatementViewMutable {
            columns: _,
            table_name: _,
            table_alias: _,
            ignore: _,
            into: _,
            or: _,
            overwrite: _,
            replace_into: _,
            table: _,
            source: _,
            after_columns,
            on: _,
            partitioned: _,
            priority,
            returning: _,
        } = insert;

        if !after_columns.is_empty() {
            return Err("not implemented: Insert::after_columns".to_string());
        }

        if priority.is_some() {
            return Err("not implemented: Insert::priority".to_string());
        }

        Ok(())
    }

    fn post_visit_insert(&mut self, insert: &mut InsertStatementViewMutable) -> VisitResult {
        let InsertStatementViewMutable {
            after_columns: _,
            columns: _,
            ignore: _,
            into: _,
            on: _,
            or: _,
            overwrite: _,
            partitioned: _,
            priority: _,
            replace_into: _,
            returning: _,
            source: _,
            table_alias: _,
            table_name,
            table: _,
        } = insert;

        let translated_db_name = convert_path_to_database(table_name, &mut self.database_names)?;

        **table_name = ObjectName(get_qualified_values_table_identifiers(&translated_db_name));

        Ok(())
    }

    fn pre_visit_ast_query(&mut self, _query: &mut Box<ast::Query>) -> VisitResult {
        self.scopes.push(Scope::new());
        Ok(())
    }

    fn post_visit_ast_query(&mut self, _query: &mut Box<ast::Query>) -> VisitResult {
        self.scopes.pop();
        Ok(())
    }

    fn pre_visit_set_operation_left(&mut self, _body: &mut SetOperationViewMutable) -> VisitResult {
        self.scopes.push(Scope::new());
        Ok(())
    }

    fn pre_visit_set_operation_right(
        &mut self,
        _body: &mut SetOperationViewMutable,
    ) -> VisitResult {
        self.scopes.pop();
        self.scopes.push(Scope::new());
        Ok(())
    }

    fn post_visit_set_operation(&mut self, _body: &mut SetOperationViewMutable) -> VisitResult {
        self.scopes.pop();
        Ok(())
    }

    fn pre_visit_table_factor_table(
        &mut self,
        relation: &mut TableFactorTableViewMut,
    ) -> VisitResult {
        let TableFactorTableViewMut {
            alias: _,
            args: _,
            name: _,
            partitions,
            version: _,
            with_hints: _,
        } = relation;

        if !partitions.is_empty() {
            return Err("not implemented: TableFactor::Table::partitions".to_string());
        }

        Ok(())
    }

    fn post_visit_table_factor_table(
        &mut self,
        relation: &mut TableFactorTableViewMut,
    ) -> VisitResult {
        let TableFactorTableViewMut {
            alias,
            args: _,
            name: table_name,
            partitions: _,
            version: _,
            with_hints: _,
        } = relation;

        let db_reference = convert_path_to_database(table_name, &mut self.database_names)?;

        let scope = self
            .scopes
            .last_mut()
            .ok_or("Expected scope to exist in traverse_table_factor")?;

        add_to_referencable_tables(
            &Some(&table_name),
            alias,
            &mut self.database_names,
            &mut scope.referencable_tables,
        )?;

        **table_name = ObjectName(get_qualified_values_table_identifiers(&db_reference));

        Ok(())
    }

    fn post_visit_table_factor_derived(
        &mut self,
        relation: &mut TableFactorDerivedViewMut,
    ) -> VisitResult {
        let TableFactorDerivedViewMut {
            alias,
            lateral: _,
            subquery: _,
        } = relation;

        let scope = self
            .scopes
            .last_mut()
            .ok_or("Expected scope to exist in traverse_table_factor_derived")?;

        add_to_referencable_tables(
            &None,
            alias,
            &mut self.database_names,
            &mut scope.referencable_tables,
        )?;

        Ok(())
    }

    fn post_visit_compound_identifier(&mut self, identifiers: &mut Vec<Ident>) -> VisitResult {
        let (table_reference, column_name) = extract_binary_identifiers(identifiers, "expression")?;

        let table_replacement_identifiers =
            get_replacement_identifiers_for_table(&self.scopes, &table_reference)?;

        let mut new_identifiers = table_replacement_identifiers.clone();
        new_identifiers.push(Ident::new(&column_name));

        *identifiers = new_identifiers;

        Ok(())
    }

    fn pre_visit_function(&mut self, func: &mut Function) -> VisitResult {
        let Function {
            args: _,
            distinct,
            filter: _,
            name,
            null_treatment,
            order_by: _,
            over: _,
            // "special" apparently means the function's parentheses are omitted
            special: _,
        } = func;

        if *distinct {
            return Err("Unhandled syntax: Function::distinct".to_string());
        }

        if null_treatment.is_some() {
            return Err("Unhandled syntax: Function::null_treatment".to_string());
        }

        validate_function_name(&name)
    }

    fn pre_visit_drop(&mut self, drop: &mut DropStatementViewMutable) -> VisitResult {
        let DropStatementViewMutable {
            cascade,
            if_exists: _,
            names: _,
            object_type,
            purge,
            restrict,
            temporary,
        } = drop;

        match object_type {
            ObjectType::Table => {}
            ObjectType::View => Err("not implemented: Statement::Drop::View".to_string())?,
            ObjectType::Index => Err("not implemented: Statement::Drop::Index".to_string())?,
            ObjectType::Schema => Err("not implemented: Statement::Drop::Schema".to_string())?,
            ObjectType::Role => Err("not implemented: Statement::Drop::Role".to_string())?,
            ObjectType::Sequence => Err("not implemented: Statement::Drop::Sequence".to_string())?,
            ObjectType::Stage => Err("not implemented: Statement::Drop::Stage".to_string())?,
        };

        if **cascade {
            Err("not implemented: Statement::Drop::cascade".to_string())?;
        };

        if **purge {
            Err("not implemented: Statement::Drop::purge".to_string())?;
        };

        if **restrict {
            Err("not implemented: Statement::Drop::restrict".to_string())?;
        };

        if **temporary {
            Err("not implemented: Statement::Drop::temporary".to_string())?;
        };

        Ok(())
    }

    fn pre_visit_on_conflict(&mut self, on_conflict: &mut OnConflict) -> VisitResult {
        let OnConflict {
            action: _,
            conflict_target,
        } = on_conflict;

        if conflict_target.is_some() {
            return Err("not implemented: OnConflict::conflict_target".to_string());
        }

        Ok(())
    }

    fn pre_visit_join_constraint(&mut self, constraint: &mut JoinConstraint) -> VisitResult {
        if let JoinConstraint::Using(_) = constraint {
            return Err("not implemented: JoinConstraint::Using".to_string());
        }

        Ok(())
    }

    fn pre_visit_function_arg_expr(
        &mut self,
        function_arg_expr: &mut FunctionArgExpr,
    ) -> VisitResult {
        if let FunctionArgExpr::QualifiedWildcard(_) = function_arg_expr {
            return Err("Unhandled syntax: FunctionArgExpr::QualifiedWildcard".to_string());
        }

        Ok(())
    }

    fn pre_visit_value(&mut self, value: &mut Value) -> VisitResult {
        match value {
            Value::Placeholder(value) => {
                if value == "?" {
                    Ok(())
                } else {
                    Err(format!("Unhandled value: Value::Placeholder({})", value))
                }
            }

            Value::SingleQuotedString(_) | Value::Number(_, _) | Value::Null => Ok(()),

            Value::DollarQuotedString(_) => {
                Err("Unhandled value: Value::DollarQuotedString".to_string())
            }
            Value::EscapedStringLiteral(_) => {
                Err("Unhandled value: Value::EscapedStringLiteral".to_string())
            }
            Value::SingleQuotedByteStringLiteral(_) => {
                Err("Unhandled value: Value::SingleQuotedByteStringLiteral".to_string())
            }
            Value::DoubleQuotedByteStringLiteral(_) => {
                Err("Unhandled value: Value::DoubleQuotedByteStringLiteral".to_string())
            }
            Value::RawStringLiteral(_) => {
                Err("Unhandled value: Value::RawStringLiteral".to_string())
            }
            Value::NationalStringLiteral(_) => {
                Err("Unhandled value: Value::NationalStringLiteral".to_string())
            }
            Value::HexStringLiteral(_) => {
                Err("Unhandled value: Value::HexStringLiteral".to_string())
            }
            Value::DoubleQuotedString(_) => {
                Err("Unhandled value: Value::DoubleQuotedString".to_string())
            }
            Value::Boolean(_) => Err("Unhandled value: Value::Boolean".to_string()),
            Value::UnQuotedString(_) => Err("Unhandled value: Value::UnQuotedString".to_string()),
        }?;

        Ok(())
    }
}

fn validate_function_name(name: &ObjectName) -> Result<(), String> {
    let ObjectName(name_identifiers) = name;

    let unary_name = extract_unary_identifier(name_identifiers, "function")?;

    let supported_functions = ["COUNT", "CURRENT_TIMESTAMP", "DATETIME", "SUM"];

    if supported_functions
        .iter()
        .any(|&func| func.eq_ignore_ascii_case(&unary_name))
    {
        Ok(())
    } else {
        Err(format!("Unsupported function: {}", unary_name))
    }
}

fn resolve_translated_db_name(databases: &mut DatabaseNamesByPath, path: &str) -> String {
    if databases.contains_key(path) {
        databases[path].clone()
    } else {
        let translated_db_name = if databases.is_empty() {
            "main".to_string()
        } else {
            format!("db{}", databases.len())
        };
        databases.insert(path.to_string(), translated_db_name.clone());
        translated_db_name
    }
}

fn convert_path_to_database(
    ObjectName(identifiers): &ObjectName,
    databases: &mut DatabaseNamesByPath,
) -> Result<String, String> {
    let path_string = extract_unary_identifier(identifiers, "table")?;

    let path_buf = Utf8UnixPathBuf::from(path_string);
    let path_buf = path_buf.normalize();

    let translated_db_name = resolve_translated_db_name(databases, &path_buf.to_string());

    Ok(translated_db_name)
}

fn get_qualified_values_table_identifiers(db_reference: &str) -> Vec<Ident> {
    vec![Ident::new(db_reference), Ident::new(VALUES_TABLE_NAME)]
}

fn add_to_referencable_tables(
    table_name: &Option<&ObjectName>,
    alias: &Option<TableAlias>,
    databases: &mut DatabaseNamesByPath,
    referencable_tables: &mut ReferencableTables,
) -> Result<(), String> {
    match (table_name, alias) {
        (
            _,
            Some(TableAlias {
                columns,
                name: alias_ident,
            }),
        ) => {
            if !columns.is_empty() {
                return Err("not implemented: TableAlias::columns".to_string());
            }

            let Ident { value, .. } = alias_ident;

            if referencable_tables.contains_key(value) {
                return Err(format!("Duplicate table alias: {}", value));
            }

            referencable_tables.insert(value.clone(), vec![alias_ident.to_owned()]);
        }
        (Some(table_name), None) => {
            let ObjectName(name_identifiers) = table_name;

            let path = extract_unary_identifier(name_identifiers, "table")?;

            let db_reference = convert_path_to_database(table_name, databases)?;

            if referencable_tables.contains_key(&path) {
                return Err(format!("Duplicate table name: {}", path));
            }

            referencable_tables.insert(path, get_qualified_values_table_identifiers(&db_reference));
        }
        (None, None) => (),
    };

    Ok(())
}

fn get_replacement_identifiers_for_table<'a>(
    scopes: &'a Vec<Scope>,
    table_reference: &str,
) -> Result<&'a Vec<Ident>, String> {
    for scope in scopes.iter().rev() {
        if let Some(replacement_identifiers) = scope.referencable_tables.get(table_reference) {
            return Ok(replacement_identifiers);
        }
    }

    Err(format!(
        "Unknown table reference: {} ({:?})",
        table_reference, scopes
    ))
}
