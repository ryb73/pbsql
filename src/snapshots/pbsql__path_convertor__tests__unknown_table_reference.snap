---
source: src/path_convertor.rs
expression: translate_result
---
Err: "Unknown table reference: ~/books/things"
